controller/rancher/certificate_fetcher.go

```diff
@@ -0,0 +1,252 @@
+	certs := []*config.Certificate{}
+
+	fetcher.mu.RLock()
+	for _, value := range fetcher.CertsCache {
+		certs = append(certs, value)
+	}
+	fetcher.mu.RUnlock()
+
+	return certs
+}
+
+func (fetcher *RCertificateFetcher) ReadDefaultCertificate(defaultCertDir string) *config.Certificate {
+	var currentDefCert *config.Certificate
+
+	fetcher.mu.RLock()
+	currentDefCert = fetcher.DefaultCert
+	fetcher.mu.RUnlock()
+
+	return currentDefCert
+}
+
+func (fetcher *RCertificateFetcher) LookForCertUpdates(certDir string, defaultCertDir string, doOnUpdate func(string)) {
+	fetcher.CertDir = certDir
+	fetcher.DefaultCertDir = defaultCertDir
+	lastUpdated := time.Now()
+	for {
+		logrus.Debugf("Start --- LookForCertUpdates polling dir %v", fetcher.CertDir)
+		forceUpdate := false
+		certsUpdatedFlag := false
+		logrus.Debugf("lastUpdated %v", lastUpdated)
+
+		if time.Since(lastUpdated).Seconds() >= fetcher.forceUpdateInterval {
+			logrus.Debugf("LookForCertUpdates: Executing force update as certs in cache have not been updated in: %v minutes", fetcher.forceUpdateInterval)
+			forceUpdate = true
+		}
+
+		//read the certs from the dir into tempMap
+		fetcher.tempCertsMap = make(map[string]*config.Certificate)
+		err := filepath.Walk(fetcher.CertDir, fetcher.readCertificate)
+		if err != nil {
+			logrus.Errorf("LookForCertUpdates: Error %v reading certs from cert dir  %v", err, fetcher.CertDir)
+		} else {
+			//compare with existing cache
+			if !reflect.DeepEqual(fetcher.CertsCache, fetcher.tempCertsMap) || forceUpdate {
+				if !forceUpdate {
+					logrus.Infof("LookForCertUpdates: Found an update in cert dir %v, updating the cache", fetcher.CertDir)
+				} else {
+					logrus.Infof("LookForCertUpdates: Force Update triggered, updating the cache from cert dir %v", fetcher.CertDir)
+				}
+				//there is some change, refresh certs
+				fetcher.mu.Lock()
+				fetcher.CertsCache = make(map[string]*config.Certificate)
+				for path, newCert := range fetcher.tempCertsMap {
+					fetcher.CertsCache[path] = newCert
+					logrus.Debugf("LookForCertUpdates: Cert is reloaded in cache : %v", newCert)
+				}
+				certsUpdatedFlag = true
+				fetcher.mu.Unlock()
+			}
+		}
+
+		//read the cert from the defaultCertDir into tempMap
+		fetcher.tempCertsMap = make(map[string]*config.Certificate)
+		err = filepath.Walk(fetcher.DefaultCertDir, fetcher.readCertificate)
+		if err != nil {
+			logrus.Errorf("LookForCertUpdates: Error %v reading default cert from dir  %v", err, fetcher.DefaultCertDir)
+		} else {
+			var tempDefCert *config.Certificate
+			for _, cert := range fetcher.tempCertsMap {
+				tempDefCert = cert
+			}
+			//compare with existing default cert
+			if !reflect.DeepEqual(fetcher.DefaultCert, tempDefCert) || forceUpdate {
+				fetcher.mu.Lock()
+				fetcher.DefaultCert = tempDefCert
+				certsUpdatedFlag = true
+				fetcher.mu.Unlock()
+			}
+		}
+
+		fetcher.mu.RLock()
```

Comment:
Do we need to lock here? No other method can change the default certificate from what it seems.